internal/convert.go

```diff
@@ -29,26 +29,26 @@ import (
-	dataSink         func(table string, cols []string, values []interface{})
-	DataFlush        func()              `json:"-"` // Data flush is used to flush out remaining writes and wait for them to complete.
-	Location         *time.Location      // Timezone (for timestamp conversion).
-	sampleBadRows    rowSamples          // Rows that generated errors during conversion.
-	Stats            stats               `json:"-"`
-	TimezoneOffset   string              // Timezone offset for timestamp conversion.
-	SpDialect        string              // The dialect of the spanner database to which Spanner migration tool is writing.
-	UniquePKey       map[string][]string // Maps Spanner table name to unique column name being used as primary key (if needed).
-	Audit            Audit               `json:"-"` // Stores the audit information for the database conversion
-	Rules            []Rule              // Stores applied rules during schema conversion
-	IsSharded        bool                // Flag denoting if the migration is sharded or not
+	mode           mode                     // Schema mode or data mode.
+	SpSchema       ddl.Schema               // Maps Spanner table name to Spanner schema.
+	SyntheticPKeys map[string]SyntheticPKey // Maps Spanner table name to synthetic primary key (if needed).
+	SrcSchema      map[string]schema.Table  // Maps source-DB table name to schema information.
+	SchemaIssues   map[string]TableIssues   // Maps source-DB table/col to list of schema conversion issues.
+	ToSpanner      map[string]NameAndCols   `json:"-"` // Maps from source-DB table name to Spanner name and column mapping.
+	ToSource       map[string]NameAndCols   `json:"-"` // Maps from Spanner table name to source-DB table name and column mapping.
+	UsedNames      map[string]bool          `json:"-"` // Map storing the names that are already assigned to tables, indices or foreign key contraints.
+	dataSink       func(table string, cols []string, values []interface{})
+	DataFlush      func()              `json:"-"` // Data flush is used to flush out remaining writes and wait for them to complete.
+	Location       *time.Location      // Timezone (for timestamp conversion).
+	sampleBadRows  rowSamples          // Rows that generated errors during conversion.
+	Stats          stats               `json:"-"`
+	TimezoneOffset string              // Timezone offset for timestamp conversion.
+	SpDialect      string              // The dialect of the spanner database to which Spanner migration tool is writing.
+	UniquePKey     map[string][]string // Maps Spanner table name to unique column name being used as primary key (if needed).
+	Audit          Audit               `json:"-"` // Stores the audit information for the database conversion
+	Rules          []Rule              // Stores applied rules during schema conversion
+	IsSharded      bool                // Flag denoting if the migration is sharded or not
+	ConvLock       sync.RWMutex        `json:"-"`
```

Comment:
In case when will support multiple migrations, this lock will need to be taken per conv object, if keep directly inside sessionstate we might need to maintain two maps in future:
1. migration id -> conv object
2. migration id -> lock

Reply:
Not inside `sessionState`, just as a variable in the code.
Example - 

```
var lock sync.RWMutex
lock.Lock()
///conv stuff
lock.Unlock()
```

Reply:
I guess what I'm trying to ask is if there is any downside to having an `RWMutex` object as a member of a struct v/s it just being a variable.

Reply:
https://stackoverflow.com/questions/68987483/using-a-mutex-within-a-struct-in-go

Reply:
A good point that this thread mentions - 

> It’s almost always a mistake to copy a sync.Mutex e.g. by passing it to another function or embedding it in a struct and making a copy of that struct.

If its inside an object like `conv`, and we create a copy of it for some reason, the mutex also gets copied and may lead to incorrect usage?

Reply:
Got confused, this makes sense. Thanks